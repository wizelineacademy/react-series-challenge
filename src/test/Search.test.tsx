import * as React from 'react';
import 'jest-enzyme';
import 'jest-styled-components'
import configureStore from 'redux-mock-store'
import { mount, shallow, render } from 'enzyme';
import SearchContainer, { Search } from '../components/Search';
import { Provider } from 'react-redux';

describe('Search', () => {
    const mockStore = configureStore();
    let store: any;

    beforeEach(() => {
        store = mockStore();
    })

    it('renders withouth crashing', () => {
        const wrapper = mount( <Search store={store} /> );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders with the given query as value', () => {
        const QUERY = 'Game of Thrones';
        store = mockStore({ search: QUERY })

        const wrapper = mount(<Provider store={store}>
            <SearchContainer />
        </Provider>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input')).toHaveValue(QUERY);
    });

    describe('when clearing the text input', () => {
        it('dispatches the fetch_gifs action', () => {
            const fetch_gifs_fn = jest.fn();
            const search_fn     = jest.fn();

            const event = { target: { value: '' } }

            const wrapper = mount(<Provider store={store}>
                <Search search={search_fn}
                    fetch_gifs={fetch_gifs_fn} />
                </Provider>)

            wrapper.find('Search').simulate('change', event);

            expect(search_fn).not.toHaveBeenCalled();
            expect(fetch_gifs_fn).toHaveBeenCalled();
        });
    });

    describe('when typing less than 3 characters', () => {
        it('does not call any action', () => {
            const fetch_gifs_fn = jest.fn();
            const search_fn     = jest.fn();

            const event = { target: { value: 'dr' } }

            const wrapper = mount(<Provider store={store}>
                <Search search={search_fn} fetch_gifs={fetch_gifs_fn} />
            </Provider>);

            wrapper.find('Search').simulate('change', event);

            expect(search_fn).not.toHaveBeenCalled();
            expect(fetch_gifs_fn).not.toHaveBeenCalled();
        });
    });

    describe('when typing more than 3 characters', () => {
        it('dispatches the search action', () => {
            const QUERY = 'dragon ball';

            const fetch_gifs_fn = jest.fn();
            const search_fn     = jest.fn();

            const event = { target: { value: QUERY } }

            const wrapper = mount(<Provider store={store}>
                <Search search={search_fn} fetch_gifs={fetch_gifs_fn} />
            </Provider>);

            wrapper.find('Search').simulate('change', event);

            expect(search_fn).toHaveBeenCalledWith(QUERY);
            expect(fetch_gifs_fn).not.toHaveBeenCalled();
        });
    });
});