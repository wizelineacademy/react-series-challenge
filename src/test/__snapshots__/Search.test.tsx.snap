// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders with the given query as value 1`] = `
.c0 {
  width: 100%;
  padding: 6px;
  margin: 0 auto;
  box-shadow: none;
  color: #333;
  box-sizing: border-box;
}

.c0:focus {
  outline: none;
}

<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Search)>
    <Search
      fetch_gifs={[Function]}
      query="Game of Thrones"
      search={[Function]}
    >
      <styled.input
        onChange={[Function]}
        placeholder="Search in trending..."
        value="Game of Thrones"
      >
        <input
          className="c0"
          onChange={[Function]}
          placeholder="Search in trending..."
          type="text"
          value="Game of Thrones"
        />
      </styled.input>
    </Search>
  </Connect(Search)>
</Provider>
`;

exports[`Search renders withouth crashing 1`] = `
.c0 {
  width: 100%;
  padding: 6px;
  margin: 0 auto;
  box-shadow: none;
  color: #333;
  box-sizing: border-box;
}

.c0:focus {
  outline: none;
}

<Search
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <styled.input
    onChange={[Function]}
    placeholder="Search in trending..."
  >
    <input
      className="c0"
      onChange={[Function]}
      placeholder="Search in trending..."
      type="text"
    />
  </styled.input>
</Search>
`;
